File: 2020m./uzd2./sprendimai/uzd2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class FishFisher {
public:
    string name;
    vector<FishFisher> fishes;
    int num;

    FishFisher(const string& name, const int num) : name(name), fishes({}), num(num) {}
};
void sortList(std::vector<FishFisher>& list) {
    int size = list.size();
    for (int i = 0; i < size; ++i) {
        for (int j = i + 1; j < size; ++j) {
            if (list[i].num < list[j].num ||
                (list[i].num == list[j].num && list[i].name > list[j].name)) {
                std::swap(list[i], list[j]);
            }
        }
    }
}

int main() {
    ifstream data("U2.txt");
    ofstream output("U2rez.txt");

    vector<FishFisher> fishers;

    int numFishers;
    data >> numFishers;
    data.ignore();

    for (int i = 0; i < numFishers; ++i) {
        string fisherData;
        getline(data, fisherData);
        string name = fisherData.substr(0, 19);
        int fishCaught = stoi(fisherData.substr(20));
        vector<FishFisher> fishes;
        for (int j = 0; j < fishCaught; ++j) {
            string fishData;
            getline(data, fishData);
            string fishName = fishData.substr(0, 19);
            int weight = stoi(fishData.substr(20));
            fishes.push_back(FishFisher(fishName, weight));
        }
        fishers.push_back(FishFisher(name, 0));
        fishers[i].fishes = fishes;
    }

    vector<FishFisher> fishes;

    int numFish;
    data >> numFish;
    data.ignore();

    for (int i = 0; i < numFish; ++i) {
        int weight = 0;
        string fishData;
        getline(data, fishData);
        string name = fishData.substr(0, 19);
        int pointWorth = stoi(fishData.substr(20));
        for (FishFisher& fisher : fishers) {
            for (FishFisher& fish : fisher.fishes) {
                if (fish.name == name) {
                    weight += fish.num;
                    fisher.num += pointWorth;
                    if (fish.num >= 200) {
                        fisher.num += 30;
                    } else {
                        fisher.num += 10;
                    }
                }
            }
        }
        fishes.push_back(FishFisher(name, weight));
    }

    sortList(fishers);
    sortList(fishes);

    output << "Dalyviai\n";

    for (const FishFisher& fisher : fishers) {
        output << fisher.name << " " << fisher.num << "\n";
    }

    output << "Laimikis\n";

    for (const FishFisher& fish : fishes) {
        output << fish.name << " " << fish.num << "\n";
    }

    data.close();
    output.close();

    return 0;
}

File: 2020m./uzd2./sprendimai/uzd2.py
class Fisher:
    score = 0
    def __init__(self, name, fishes):
        self.name = name 
        self.fishes = fishes

class Fish:
    def __init__(self, name, weight):
        self.name = name
        self.weight = weight

def sortList(list, values):
    size = len(list)
    for i in range(size):
        for j in range(i + 1, size):
            if(values[i] < values[j] or 
                (values[i] == values[j]
                and list[i].name > list[j].name)):
                list[i], list[j] = list[j], list[i]
                values[j], values[i] = values[i], values[j]
    return list

def main():
    data = open("U2.txt", "r")
    output = open("U2rez.txt", "w")

    fishers = []

    for i in range(int(next(data))):
        fisherData = next(data)
        name = fisherData[:19]
        fishCaught = int(fisherData[20:])
        fishes = []
        for j in range(fishCaught):
            fishData = next(data)
            fishName = fishData[:19]
            weight = int(fishData[20:])
            fishes.append(Fish(fishName, weight))
        fishers.append(Fisher(name, fishes))

    fishes = []
    
    for i in range(int(next(data))):
        weight = 0
        fishData = next(data)
        name = fishData[:19]
        pointWorth = int(fishData[20:])
        for fisher in fishers:
            for fish in fisher.fishes:
                if(fish.name == name):
                    weight += fish.weight
                    fisher.score += pointWorth
                    if(fish.weight >= 200):
                        fisher.score += 30
                    else:
                        fisher.score += 10
        fishes.append(Fish(name, weight))

    fishers = sortList(fishers, [fisher.score for fisher in fishers])
    fishes = sortList(fishes, [fish.weight for fish in fishes])

    output.write("Dalyviai\n")

    for fisher in fishers:
        output.write(fisher.name + " " + str(fisher.score) + "\n")

    output.write("Laimikis\n")

    for fish in fishes:
        output.write(fish.name + " " + str(fish.weight) + "\n")
    
    data.close()
    output.close()

if __name__ == "__main__":
    main()